transaction/core/src/amount/v2.rs

```diff
@@ -0,0 +1,341 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+//! A commitment to an output's amount, denominated in picoMOB.
+//!
+//! Amounts are implemented as Pedersen commitments. The associated private keys
+//! are "masked" using a shared secret.
+
+use crate::{
+    domain_separators::{
+        AMOUNT_BLINDING_DOMAIN_TAG, AMOUNT_SHARED_SECRET_DOMAIN_TAG, AMOUNT_TOKEN_ID_DOMAIN_TAG,
+        AMOUNT_VALUE_DOMAIN_TAG,
+    },
+    Amount, AmountError, TokenId,
+};
+use alloc::vec::Vec;
+use core::convert::TryInto;
+use crc::Crc;
+use hkdf::Hkdf;
+use mc_crypto_digestible::Digestible;
+use mc_crypto_hashes::{Blake2b512, Digest};
+use mc_crypto_keys::RistrettoPublic;
+use mc_crypto_ring_signature::{generators, CompressedCommitment, Scalar};
+use prost::Message;
+use serde::{Deserialize, Serialize};
+use sha2::Sha512;
+use zeroize::Zeroize;
+
+/// A commitment to an amount of MobileCoin or a related token, as it appears on
+/// the blockchain. This is a "blinded" commitment, and only the sender and
+/// receiver know the value and token id.
+///
+/// This differs from MaskedAmountV1 in that there is an "amount shared secret"
+/// which is computed by hashing from "tx out shared secret". The purpose of
+/// this is to make it possible to selectively reveal the amount and token id of
+/// a TxOut, without revealing the memo and other things. (See also MCIP #42).
+#[derive(Clone, Deserialize, Digestible, Eq, Hash, Message, PartialEq, Serialize, Zeroize)]
+pub struct MaskedAmountV2 {
+    /// A Pedersen commitment `v*H + b*G` to a quantity `v` of MobileCoin or a
+    /// related token, with blinding `b`,
+    #[prost(message, required, tag = "1")]
+    pub commitment: CompressedCommitment,
+
+    /// `masked_value = value XOR_8 Blake2B(value_mask | shared_secret)`
+    #[prost(fixed64, required, tag = "2")]
+    pub masked_value: u64,
+
+    /// `masked_token_id = token_id XOR_8 Blake2B(token_id_mask |
+    /// shared_secret)` 8 bytes long when used, 0 bytes for older amounts
+    /// that don't have this.
+    #[prost(bytes, tag = "3")]
+    pub masked_token_id: Vec<u8>,
+}
+
+impl MaskedAmountV2 {
+    /// Creates a commitment `value*H + blinding*G`, and "masks" the commitment
+    /// secrets so that they can be recovered by the recipient.
+    ///
+    /// # Arguments
+    /// * `amount` - The amount information to be masked
+    /// * `shared_secret` - The shared secret, e.g. `rB` for transaction private
+    ///   key `r` and recipient public key `B`.
+    #[inline]
+    pub fn new(
+        amount: Amount,
+        tx_out_shared_secret: &RistrettoPublic,
+    ) -> Result<Self, AmountError> {
+        let amount_shared_secret = get_amount_shared_secret(tx_out_shared_secret);
```

Comment:
we shipped the original blockchain using hkdf blake2, it will break all the existing TxOuts if we try to change that for V1. In V2 we can do it correctly

Reply:
i see -- so the question is, Amount Shared Secret is new, so shouldn't it use SHA512 instead of blake2b here?

I think you are absolutely right, and I meant to change this. Thanks for spotting this